src/archive/zip.rs

```diff
@@ -105,11 +119,20 @@ where
         }
     }
 
+    let password = password.map(|p| p.to_owned());
+
     let (tx, rx) = mpsc::channel();
     thread::spawn(move || {
         for idx in 0..archive.len() {
             let maybe_file_in_archive = (|| {
-                let file = match archive.by_index(idx) {
+                let zip_result = match password.clone() {
+                    Some(password) => archive
+                        .by_index_decrypt(idx, password.as_bytes()).ok()?
+                        .map_err(|_| zip::result::ZipError::UnsupportedArchive("Password required to decrypt file")),
```

Comment:
Ah, now I see the problem, this closure is returning `Option<Result<_>>` instead of `Result<Option<_>>`, that's why you couldn't do `??` and just return the error.

Well, I guess there is no reason for us to here return `Option`, my bet is that `by_index` doesn't fail because `idx` is always in bounds.

Anyways, with `.ok()?` we do lose context on why this failed, we need `?` without the `.ok()?`, can you try making this closure return `Result<FileInArchive>` instead of `Option<Result<FileInArchive>>`?

Reply:
yes, this is annoying.  
because:

```rs
    pub fn by_index_decrypt<'a>(
        &'a mut self,
        file_number: usize,
        password: &[u8],
    ) -> ZipResult<Result<ZipFile<'a>, InvalidPassword>> {
        self.by_index_with_optional_password(file_number, Some(password))
    }

    /// Get a contained file by index
    pub fn by_index(&mut self, file_number: usize) -> ZipResult<ZipFile<'_>> {
        Ok(self
            .by_index_with_optional_password(file_number, None)?
            .unwrap())
    }
```

but we can not call `by_index_with_optional_password` because it is not public exported.

Reply:
done in 164f800

Reply:
Oh I see what you mean now, yes that is annoying.

Code is good tho ty.